File: frontend/src/api/chatStreamProtocol.ts
import type { MessageRoute, PromptApproval, RegenerationSnapshot, RouteChoice } from "../types/chat";

export const CHAT_STREAM_PROTOCOL = "laf-chat-stream" as const;
export const CHAT_STREAM_VERSION = 1 as const;

export interface InspirationImage {
  thumb_url: string;
  full_url: string;
  source_url: string;
  width?: number;
  height?: number;
  title?: string;
}

export interface StreamInspirationCard {
  id?: string;
  title: string;
  content: string;
  sources: { title: string; url: string }[];
  images?: InspirationImage[];
  selected?: string[];
}

export interface IllustrationSceneSpec {
  narrative: string;
  protected_narrative?: string;
  repo_id?: string;
  thread_id?: string;
  turn_id?: string;
  draft_prompt: string;
  appearance?: string;
  wardrobe: string;
  locale: string;
  actors: string[];
  subjects?: Array<{ name: string; description?: string; weight?: number }>;
  rating: "sfw" | "nsfw";
  character_lora?: boolean;
  profile?: string;
  profile_prompt?: string;
  negative_prompt?: string;
  aspect_ratio?: "1:1" | "2:3" | "3:2" | "3:4" | "4:3" | "9:16" | "16:9";
  camera?: string;
  composition?: string;
}

export interface AudioDialogueLine {
  speaker: string;
  text: string;
  emotion?: Record<string, number>;
}

/** V1.5 默认开放：结构化视频参数（后端 dry-run 组装结果，供测试核对参数是否上传） */
export interface VideoParams {
  mode: "climax" | "firstlast" | "transition";
  model: string;
  size: string;
  endpoint: string;
  images: string[];
  reference_binding: Record<string, string>;
  warnings: string[];
}

export type ChatStreamEvent =
  | { type: "trace"; text: string }
  | { type: "delta"; text: string }
  | { type: "thinking"; text: string }
  | { type: "replace"; text: string }
  | { type: "route"; route: MessageRoute }
  | { type: "image"; url: string; id?: string; regeneration?: RegenerationSnapshot }
  | { type: "video"; url: string; id?: string }
  | { type: "illustrate_request"; prompt: string; motion: number; actors: string[]; sceneSpec?: IllustrationSceneSpec; id?: string; offset?: number; turnId?: string; videoMode?: "climax" | "firstlast"; firstFrameDesc?: string; lastFrameDesc?: string; prevTailDesc?: string; lastFrameUrl?: string; transition?: "reuse" | "regenerate" | "ambiguous"; videoPrompt?: string; videoParams?: VideoParams; transitionVideoPrompt?: string; transitionVideoParams?: VideoParams }
  | { type: "audio_request"; lines: AudioDialogueLine[]; id?: string }
  | { type: "rag_status"; state: string; kind: string; count?: number }
  | { type: "inspiration"; card: StreamInspirationCard }
  | { type: "approval"; approval: PromptApproval }
  | { type: "route_choice"; choice: RouteChoice }
  | { type: "interrupted" }
  | { type: "error"; message: string };

function record(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`对话流协议错误：${label} 不是对象`);
  }
  return value as Record<string, unknown>;
}

function requiredString(data: Record<string, unknown>, key: string): string {
  if (typeof data[key] !== "string") {
    throw new Error(`对话流协议错误：缺少字符串字段 data.${key}`);
  }
  return data[key];
}

// 灵感卡字段归一化：新结构 {title, content, sources, images?, selected?} 为准；兼容旧结构 {query, prompt, tags, sources}。
export function normalizeInspirationCard(card: Record<string, unknown>): StreamInspirationCard {
  const title = typeof card.title === "string" && card.title
    ? card.title
    : typeof card.query === "string" ? card.query : "";
  const content = typeof card.content === "string" && card.content
    ? card.content
    : typeof card.prompt === "string" ? card.prompt : "";
  const sources = Array.isArray(card.sources)
    ? card.sources as { title: string; url: string }[]
    : [];
  const images = Array.isArray(card.images)
    ? card.images as StreamInspirationCard["images"]
    : [];
  const selected = Array.isArray(card.selected)
    ? card.selected.map((u) => String(u))
    : [];
  return { ...(typeof card.id === "string" ? { id: card.id } : {}), title, content, sources, images, selected };
}

export function decodeChatStreamEvent(value: unknown): ChatStreamEvent {
  const envelope = record(value, "事件");
  if (envelope.protocol !== CHAT_STREAM_PROTOCOL || envelope.version !== CHAT_STREAM_VERSION) {
    throw new Error(
      `不支持的对话流协议：${String(envelope.protocol || "unknown")} v${String(envelope.version ?? "?")}`,
    );
  }
  const data = record(envelope.data, "data");
  switch (envelope.type) {
    case "trace":
      return { type: "trace", text: requiredString(data, "text") };
    case "delta":
      return { type: "delta", text: requiredString(data, "text") };
    case "thinking":
      // 思考全公开（2026-08-31 晚）：think 流式增量进独立思考面板，不混入正文。
      return { type: "thinking", text: requiredString(data, "text") };
    case "replace":
      return { type: "replace", text: requiredString(data, "text") };
    case "route":
      return { type: "route", route: requiredString(data, "route") as MessageRoute };
    case "image":
      return {
        type: "image",
        url: requiredString(data, "url"),
        ...(typeof data.id === "string" ? { id: data.id } : {}),
        ...(data.regeneration ? { regeneration: data.regeneration as RegenerationSnapshot } : {}),
      };
    case "video":
      return {
        type: "video",
        url: requiredString(data, "url"),
        ...(typeof data.id === "string" ? { id: data.id } : {}),
      };
    case "illustrate_request":
      return {
        type: "illustrate_request",
        prompt: requiredString(data, "prompt"),
        motion: typeof data.motion === "number" ? data.motion : 0,
        actors: Array.isArray(data.actors) ? data.actors.map((a) => String(a)) : [],
        ...(data.scene_spec && typeof data.scene_spec === "object"
          ? { sceneSpec: data.scene_spec as IllustrationSceneSpec }
          : {}),
        ...(typeof data.id === "string" ? { id: data.id } : {}),
        ...(typeof data.offset === "number" ? { offset: data.offset } : {}),
        ...(typeof data.turn_id === "string" ? { turnId: data.turn_id } : {}),
        // V1.5/B1 视频协议可选字段：宽松透传（旧后端不带 → 无这些字段）
        ...(data.video_mode === "climax" || data.video_mode === "firstlast"
          ? { videoMode: data.video_mode }
          : {}),
        ...(typeof data.first_frame_desc === "string" ? { firstFrameDesc: data.first_frame_desc } : {}),
        ...(typeof data.last_frame_desc === "string" ? { lastFrameDesc: data.last_frame_desc } : {}),
        ...(typeof data.prev_tail_desc === "string" ? { prevTailDesc: data.prev_tail_desc } : {}),
        ...(typeof data.last_frame_url === "string" ? { lastFrameUrl: data.last_frame_url } : {}),
        // V1.5/W2：首帧复用决策合并结果（三态）宽松解码（旧后端不带 → 无此字段）
        ...(data.transition === "reuse" || data.transition === "regenerate" || data.transition === "ambiguous"
          ? { transition: data.transition }
          : {}),
        // V1.5 默认开放：climax 视频提示词随事件下发（无视频模板也生成，供测试核对）
        ...(typeof data.video_prompt === "string" ? { videoPrompt: data.video_prompt } : {}),
        // V1.5 默认开放：结构化视频参数（dry-run 组装结果，供测试核对参数是否上传）
        ...(data.video_params && typeof data.video_params === "object"
          ? { videoParams: data.video_params as VideoParams }
          : {}),
        // W3 转场视频：转场提示词 + 参数（firstlast + transition≠reuse 时后端下发）
        ...(typeof data.transition_video_prompt === "string"
          ? { transitionVideoPrompt: data.transition_video_prompt }
          : {}),
        ...(data.transition_video_params && typeof data.transition_video_params === "object"
          ? { transitionVideoParams: data.transition_video_params as VideoParams }
          : {}),
      };
    case "audio_request":
      return {
        type: "audio_request",
        lines: Array.isArray(data.lines)
          ? data.lines
              .map((item) => {
                if (!item || typeof item !== "object") return null;
                const line = item as Record<string, unknown>;
                const speaker = typeof line.speaker === "string" ? line.speaker : "";
                const text = typeof line.text === "string" ? line.text : "";
                if (!speaker || !text) return null;
                return {
                  speaker,
                  text,
                  ...(line.emotion && typeof line.emotion === "object"
                    ? { emotion: line.emotion as Record<string, number> }
                    : {}),
                };
              })
              .filter((line): line is AudioDialogueLine => line !== null)
          : [],
        ...(typeof data.id === "string" ? { id: data.id } : {}),
      };
    case "rag_status":
      return {
        type: "rag_status",
        state: requiredString(data, "state"),
        kind: typeof data.kind === "string" ? data.kind : "",
        ...(typeof data.count === "number" ? { count: data.count } : {}),
      };
    case "inspiration":
      return { type: "inspiration", card: normalizeInspirationCard(record(data.card, "data.card")) };
    case "approval":
      return { type: "approval", approval: record(data.approval, "data.approval") as unknown as PromptApproval };
    case "route_choice":
      return { type: "route_choice", choice: record(data.choice, "data.choice") as unknown as RouteChoice };
    case "interrupted":
      return { type: "interrupted" };
    case "error":
      return { type: "error", message: requiredString(data, "message") };
    default:
      throw new Error(`不支持的对话流事件：${String(envelope.type)}`);
  }
}

File: backend/app/services/chat_stream_protocol.py
"""对话流事件协议 v1：把内部领域事件编码成稳定的 SSE wire 结构。"""
from __future__ import annotations

from typing import Literal, Mapping, TypedDict


PROTOCOL = "laf-chat-stream"
VERSION = 1


class ChatStreamEvent(TypedDict):
    protocol: Literal["laf-chat-stream"]
    version: Literal[1]
    type: str
    data: dict


def _wire(event_type: str, data: dict) -> ChatStreamEvent:
    return {
        "protocol": PROTOCOL,
        "version": VERSION,
        "type": event_type,
        "data": data,
    }


def error_event(message: str) -> ChatStreamEvent:
    return _wire("error", {"message": message})


def encode_event(event: Mapping[str, object]) -> ChatStreamEvent | None:
    """编码一个内部事件；完成信号由 SSE 传输层收尾，不重复进入 payload。"""
    signals: list[str] = []
    for key in ("trace", "delta", "replace", "route", "image", "video", "illustrate_request", "audio_request",
                "insp", "rag_status", "approval", "route_choice", "error"):
        if key in event and event[key] is not None:
            signals.append(key)
    if event.get("interrupted") is True:
        signals.append("interrupted")
    if event.get("done") is True:
        signals.append("done")

    if signals == ["done"]:
        return None
    if len(signals) != 1:
        raise ValueError(f"对话流内部事件必须且只能包含一种事件类型：{signals or list(event)}")

    kind = signals[0]
    if kind in ("trace", "delta", "replace"):
        return _wire(kind, {"text": str(event[kind])})
    if kind == "route":
        return _wire("route", {"route": str(event["route"])})
    if kind in ("image", "video"):
        data = {"url": str(event[kind])}
        event_id = event.get("id") or event.get("image_id")
        if event_id:
            data["id"] = str(event_id)
        if kind == "image" and event.get("regeneration") is not None:
            data["regeneration"] = event["regeneration"]
        return _wire(kind, data)
    if kind == "illustrate_request":
        req = event["illustrate_request"]
        prompt = req.get("prompt", "") if isinstance(req, Mapping) else ""
        motion = req.get("motion", 0) if isinstance(req, Mapping) else 0
        actors = req.get("actors", []) if isinstance(req, Mapping) else []
        data: dict = {"prompt": str(prompt), "motion": int(motion) if isinstance(motion, (int, float)) else 0,
                      "actors": [str(a) for a in actors] if isinstance(actors, list) else []}
        if isinstance(req, Mapping) and isinstance(req.get("scene_spec"), Mapping):
            data["scene_spec"] = dict(req["scene_spec"])
        if isinstance(req, Mapping) and isinstance(req.get("offset"), (int, float)):
            data["offset"] = max(0, int(req["offset"]))
        if isinstance(req, Mapping) and req.get("turn_id"):
            data["turn_id"] = str(req["turn_id"])
        # V1.5/B1：透传视频协议可选字段（字符串且非空才带，旧数据/旧前端宽松忽略）
        if isinstance(req, Mapping):
            for _field in ("video_mode", "first_frame_desc", "last_frame_desc",
                           "prev_tail_desc", "last_frame_url", "video_prompt",
                           "transition"):
                _value = req.get(_field)
                if isinstance(_value, str) and _value:
                    data[_field] = _value
        # V1.5 默认开放：透传结构化视频参数（dry-run 组装结果，供测试核对参数是否上传）
        if isinstance(req, Mapping) and isinstance(req.get("video_params"), Mapping):
            data["video_params"] = dict(req["video_params"])
        if event.get("id"):
            data["id"] = str(event["id"])
        return _wire("illustrate_request", data)
    if kind == "audio_request":
        req = event["audio_request"]
        lines: list[dict] = []
        if isinstance(req, Mapping) and isinstance(req.get("lines"), list):
            for item in req["lines"]:
                if not isinstance(item, Mapping):
                    continue
                speaker = str(item.get("speaker") or "").strip()
                text = str(item.get("text") or "").strip()
                if not speaker or not text:
                    continue
                emotion = item.get("emotion")
                line: dict = {"speaker": speaker, "text": text}
                if isinstance(emotion, Mapping):
                    line["emotion"] = {
                        key: max(0.0, min(1.0, float(emotion[key])))
                        for key in ("happy", "angry", "sad", "fear", "hate", "low", "surprise", "neutral")
                        if isinstance(emotion.get(key), (int, float))
                    }
                lines.append(line)
        data: dict = {"lines": lines}
        if event.get("id"):
            data["id"] = str(event["id"])
        return _wire("audio_request", data)
    if kind == "insp":
        return _wire("inspiration", {"card": event["insp"]})
    if kind == "rag_status":
        rs = event["rag_status"]
        rs = rs if isinstance(rs, Mapping) else {}
        data = {"state": str(rs.get("state", "")), "kind": str(rs.get("kind", ""))}
        if rs.get("count") is not None:
            data["count"] = int(rs["count"]) if isinstance(rs["count"], (int, float)) else 0
        return _wire("rag_status", data)
    if kind == "approval":
        return _wire("approval", {"approval": event["approval"]})
    if kind == "route_choice":
        return _wire("route_choice", {"choice": event["route_choice"]})
    if kind == "interrupted":
        return _wire("interrupted", {})
    return error_event(str(event["error"]))
